### PR TITLE
Alerting: validate mute timings in the alertmanager configuration

### PR DESCRIPTION
**Pull request by JohnnyQQQQ**
**What this PR does / why we need it**: We already ensure that mute timing names are unique one runtime but should also check uniqueness when updating configs through the API.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

The validation code is taken from the upstream alertmanager, so the behavior is the same.

### CODE REVIEW

**Review by gotjosh — COMMENTED** [review]
Almost all of the validation is made at marshalling as opposed to the API level. This is to ensure that if you change the payload through the database manually, you still get this validation - can we make sure we validate it there?

There's other validation I don't see here - take a look at https://github.com/prometheus/alertmanager/blob/main/config/config.go so that we can ensure it is validated in the same way.

File: pkg/services/ngalert/api/tooling/definitions/alertmanager.go [hunk 1]
```diff
@@ -746,13 +746,43 @@ func (c *Config) UnmarshalJSON(b []byte) error {
...
 	for _, r := range c.InhibitRules {
 		if err := r.UnmarshalYAML(noopUnmarshal); err != nil {
 			return err
 		}
 	}
 
+	tiNames := make(map[string]struct{})
+	for _, mt := range c.MuteTimeIntervals {
+		if mt.Name == "" {
+			return fmt.Errorf("missing name in mute time interval")
```
Comment (gerobinson) [inline pkg/services/ngalert/api/tooling/definitions/alertmanager.go:762]: Perhaps we should include the index in the error message so it's easier to find the invalid mute timing?

```suggestion
			return fmt.Errorf("missing name in mute time interval at index %d", ix)
```

File: pkg/services/ngalert/api/tooling/definitions/alertmanager.go [hunk 2]
```diff
@@ -746,13 +746,43 @@ func (c *Config) UnmarshalJSON(b []byte) error {
...
+	return checkTimeInterval(c.Route, tiNames)
+}
+
+func checkTimeInterval(r *Route, timeIntervals map[string]struct{}) error {
+	for _, sr := range r.Routes {
+		if err := checkTimeInterval(sr, timeIntervals); err != nil {
+			return err
+		}
+	}
+	if len(r.MuteTimeIntervals) == 0 {
```
Comment (gerobinson) [inline pkg/services/ngalert/api/tooling/definitions/alertmanager.go:778]: I don't think this is required as when `len(r.MuteTimeIntervals) == 0` then `range` does nothing, but I suppose it's nice to be explicit?

**Review by grobinson-grafana — APPROVED** [review]

**Review by gotjosh — APPROVED** [review]
LGTM

File: pkg/services/ngalert/api/tooling/definitions/alertmanager.go (hunk 2 above)
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/api/tooling/definitions/alertmanager.go:778]: Yeah, I took it from upstream, I think they wanted to be explicit about it.

File: pkg/services/ngalert/api/tooling/definitions/alertmanager_test.go [hunk 3]
```diff
@@ -405,6 +405,292 @@ email_configs:
...
 }
 
+func Test_ConfigUnmashaling(t *testing.T) {
+	for _, tc := range []struct {
+		desc, input, errMsg string
+		err                 bool
+	}{
+		{
+			desc:   "empty mute time name should error",
+			err:    true,
```
[the remark this answers is not in the document]
Comment (gerobinson) [reply pkg/services/ngalert/api/tooling/definitions/alertmanager_test.go:415]: Didn't see this, but Josh is right :)

File: pkg/services/ngalert/api/tooling/definitions/alertmanager.go (hunk 1 above)
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/api/tooling/definitions/alertmanager.go:762]: This is sensible, I'll change it 👍

File: pkg/services/ngalert/api/tooling/definitions/alertmanager_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/api/tooling/definitions/alertmanager_test.go:415]: I had to use `require.Equals` as one test is valid and would return `nil`. But it works fine 👍

File: pkg/services/ngalert/api/tooling/definitions/alertmanager.go (hunk 1 above)
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/api/tooling/definitions/alertmanager.go:762]: I just thought about this. As we want to move those changes upstream, the error message would then change back again. So I will stick to as it is.

**Pull request merged**